In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt

In [2]:
df = pd.read_csv("../data/raw/Telco-Customer-Churn.csv")
df.head()

,customerID,gender,SeniorCitizen,Partner,Dependents,tenure,PhoneService,MultipleLines,InternetService,OnlineSecurity,...,DeviceProtection,TechSupport,StreamingTV,StreamingMovies,Contract,PaperlessBilling,PaymentMethod,MonthlyCharges,TotalCharges,Churn
0,7590-VHVEG,Female,0,Yes,No,1,No,No phone service,DSL,No,...,No,No,No,No,Month-to-month,Yes,Electronic check,29.85,29.85,No
1,5575-GNVDE,Male,0,No,No,34,Yes,No,DSL,Yes,...,Yes,No,No,No,One year,No,Mailed check,56.95,1889.5,No
2,3668-QPYBK,Male,0,No,No,2,Yes,No,DSL,Yes,...,No,No,No,No,Month-to-month,Yes,Mailed check,53.85,108.15,Yes
3,7795-CFOCW,Male,0,No,No,45,No,No phone service,DSL,Yes,...,Yes,Yes,No,No,One year,No,Bank transfer (automatic),42.30,1840.75,No
4,9237-HQITU,Female,0,No,No,2,Yes,No,Fiber optic,No,...,No,No,No,No,Month-to-month,Yes,Electronic check,70.70,151.65,Yes


In [3]:
df.shape

(7043, 21)

In [4]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 7043 entries, 0 to 7042
Data columns (total 21 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   customerID        7043 non-null   object 
 1   gender            7043 non-null   object 
 2   SeniorCitizen     7043 non-null   int64  
 3   Partner           7043 non-null   object 
 4   Dependents        7043 non-null   object 
 5   tenure            7043 non-null   int64  
 6   PhoneService      7043 non-null   object 
 7   MultipleLines     7043 non-null   object 
 8   InternetService   7043 non-null   object 
 9   OnlineSecurity    7043 non-null   object 
 10  OnlineBackup      7043 non-null   object 
 11  DeviceProtection  7043 non-null   object 
 12  TechSupport       7043 non-null   object 
 13  StreamingTV       7043 non-null   object 
 14  StreamingMovies   7043 non-null   object 
 15  Contract          7043 non-null   object 
 16  PaperlessBilling  7043 non-null   object 


In [5]:
df.describe(include="all")

,customerID,gender,SeniorCitizen,Partner,Dependents,tenure,PhoneService,MultipleLines,InternetService,OnlineSecurity,...,DeviceProtection,TechSupport,StreamingTV,StreamingMovies,Contract,PaperlessBilling,PaymentMethod,MonthlyCharges,TotalCharges,Churn
count,7043,7043,7043.000000,7043,7043,7043.000000,7043,7043,7043,7043,...,7043,7043,7043,7043,7043,7043,7043,7043.000000,7043,7043
unique,7043,2,NaN,2,2,NaN,2,3,3,3,...,3,3,3,3,3,2,4,NaN,6531,2
top,7590-VHVEG,Male,NaN,No,No,NaN,Yes,No,Fiber optic,No,...,No,No,No,No,Month-to-month,Yes,Electronic check,NaN,,No
freq,1,3555,NaN,3641,4933,NaN,6361,3390,3096,3498,...,3095,3473,2810,2785,3875,4171,2365,NaN,11,5174
mean,NaN,NaN,0.162147,NaN,NaN,32.371149,NaN,NaN,NaN,NaN,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,64.761692,NaN,NaN
std,NaN,NaN,0.368612,NaN,NaN,24.559481,NaN,NaN,NaN,NaN,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,30.090047,NaN,NaN
min,NaN,NaN,0.000000,NaN,NaN,0.000000,NaN,NaN,NaN,NaN,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,18.250000,NaN,NaN
25%,NaN,NaN,0.000000,NaN,NaN,9.000000,NaN,NaN,NaN,NaN,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,35.500000,NaN,NaN
50%,NaN,NaN,0.000000,NaN,NaN,29.000000,NaN,NaN,NaN,NaN,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,70.350000,NaN,NaN
75%,NaN,NaN,0.000000,NaN,NaN,55.000000,NaN,NaN,NaN,NaN,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,89.850000,NaN,NaN


In [6]:
df.columns

Index(['customerID', 'gender', 'SeniorCitizen', 'Partner', 'Dependents',
       'tenure', 'PhoneService', 'MultipleLines', 'InternetService',
       'OnlineSecurity', 'OnlineBackup', 'DeviceProtection', 'TechSupport',
       'StreamingTV', 'StreamingMovies', 'Contract', 'PaperlessBilling',
       'PaymentMethod', 'MonthlyCharges', 'TotalCharges', 'Churn'],
      dtype='object')

In [7]:
df.nunique()

customerID          7043
gender                 2
SeniorCitizen          2
Partner                2
Dependents             2
tenure                73
PhoneService           2
MultipleLines          3
InternetService        3
OnlineSecurity         3
OnlineBackup           3
DeviceProtection       3
TechSupport            3
StreamingTV            3
StreamingMovies        3
Contract               3
PaperlessBilling       2
PaymentMethod          4
MonthlyCharges      1585
TotalCharges        6531
Churn                  2
dtype: int64

## Data Dictionary

| Column | Description | Category |
|---|---|---|
| customerID | Unique identifier for each customer | Identifier |
| gender | Customer gender | Demographic |
| SeniorCitizen | Whether the customer is a senior citizen | Demographic |
| Partner | Whether the customer has a partner | Demographic |
| Dependents | Whether the customer has dependents | Demographic |
| tenure | Number of months the customer has stayed | Customer profile |
| PhoneService | Whether the customer has phone service | Service |
| MultipleLines | Whether the customer has multiple phone lines | Service |
| InternetService | Type of internet service | Service |
| OnlineSecurity | Whether online security is subscribed to | Service |
| OnlineBackup | Whether online backup is subscribed to | Service |
| DeviceProtection | Whether device protection is subscribed to | Service |
| TechSupport | Whether technical support is subscribed to | Service |
| StreamingTV | Whether the customer subscribes to streaming TV | Service |
| StreamingMovies | Whether the customer subscribes to streaming movies | Service |
| Contract | Customer's contract type | Contract |
| PaperlessBilling | Whether the customer uses paperless billing | Billing |
| PaymentMethod | Customer's payment method | Billing |
| MonthlyCharges | Monthly amount charged to the customer | Financial |
| TotalCharges | Total amount charged to the customer | Financial |
| Churn | Whether the customer left the company | Target |

### Our business problem
What factors are associated with customer churn, and which customer segments should the company prioritize for retention?

In [8]:
df["customerID"].duplicated().sum()

np.int64(0)

In [9]:
df.isna().sum()

customerID          0
gender              0
SeniorCitizen       0
Partner             0
Dependents          0
tenure              0
PhoneService        0
MultipleLines       0
InternetService     0
OnlineSecurity      0
OnlineBackup        0
DeviceProtection    0
TechSupport         0
StreamingTV         0
StreamingMovies     0
Contract            0
PaperlessBilling    0
PaymentMethod       0
MonthlyCharges      0
TotalCharges        0
Churn               0
dtype: int64

TotalCharges is suspicious, We previously saw: TotalCharges = object even though it represents a numerical amount. And now we know:
df.isna().sum() → TotalCharges = 0

This tells us that there aren't actual NaN values, but it doesn't prove that every TotalCharges value is valid numerical data. The column could contain blank spaces let's test that:

In [10]:
df["TotalCharges"].unique()

array(['29.85', '1889.5', '108.15', ..., '346.45', '306.6', '6844.5'],
      dtype=object)

In [11]:
pd.to_numeric(df["TotalCharges"], errors="coerce").isna().sum()

np.int64(11)

np.int64(11) means there are 11 values in TotalCharges that cannot be interpreted as numbers. And this explains why:
df.isna().sum() gave: TotalCharges    0

In [12]:
#lets investigate the 11 values
df[pd.to_numeric(df["TotalCharges"], errors="coerce").isna()][
    ["customerID", "tenure", "MonthlyCharges", "TotalCharges"]
]

,customerID,tenure,MonthlyCharges,TotalCharges
488,4472-LVYGI,0,52.55,
753,3115-CZMZD,0,20.25,
936,5709-LVOEQ,0,80.85,
1082,4367-NUYAO,0,25.75,
1340,1371-DWPAZ,0,56.05,
3331,7644-OMVMY,0,19.85,
3826,3213-VVOLG,0,25.35,
4380,2520-SGTTA,0,20.00,
5218,2923-ARZLG,0,19.70,
6670,4075-WKNIU,0,73.35,


All 11 problematic records have tenure = 0.
And their MonthlyCharges are valid: These are customers who have just joined, so they have no accumulated charges yet.

Therefore, for our analysis,
A blank TotalCharges for a customer with tenure = 0 can reasonably be treated as 0

### Clean TotalCharges

In [13]:
#Convert the column to numeric
df["TotalCharges"] = pd.to_numeric(df["TotalCharges"], errors="coerce")

In [14]:
#Replace the 11 invalid values with 0
df["TotalCharges"] = df["TotalCharges"].fillna(0)

In [15]:
df["TotalCharges"].isna().sum()

np.int64(0)

In [16]:
df["TotalCharges"].dtype

dtype('float64')

### Data Quality Audit

In [17]:
# 1. Dataset dimensions
df.shape

(7043, 21)

In [18]:
# 2. Missing values
df.isna().sum()

customerID          0
gender              0
SeniorCitizen       0
Partner             0
Dependents          0
tenure              0
PhoneService        0
MultipleLines       0
InternetService     0
OnlineSecurity      0
OnlineBackup        0
DeviceProtection    0
TechSupport         0
StreamingTV         0
StreamingMovies     0
Contract            0
PaperlessBilling    0
PaymentMethod       0
MonthlyCharges      0
TotalCharges        0
Churn               0
dtype: int64

In [19]:
# 3. Duplicate rows
df.duplicated().sum()

np.int64(0)

### QUESTION ONE
How big is the churn problem?

In [20]:
df["Churn"].value_counts()

Churn
No     5174
Yes    1869
Name: count, dtype: int64

In [21]:
df["Churn"].value_counts(normalize=True) * 100

Churn
No     73.463013
Yes    26.536987
Name: proportion, dtype: float64

26.54% churn rate. That's roughly 1 in every 4 customers leaving.

A 26.54% churn rate is substantial. It means the company is losing more than a quarter of its customer base over the period represented by the dataset.

In [22]:
#Find the biggest churn differences. investigate the variables most likely to tell a business story.

### Contract and churn
To investigate customer commitment.

In [28]:
contract_churn=pd.crosstab(
    df["Contract"],
    df["Churn"],
    normalize="index"
) * 100

contract_churn

Churn,No,Yes
Contract,,
Month-to-month,57.290323,42.709677
One year,88.730482,11.269518
Two year,97.168142,2.831858


Month-to-month customers are substantially more likely to churn, while two-year contract customers have the lowest churn rate. it suggests contract commitment is strongly associated with customer retention.

In [29]:
#To calclate the exact gap beteen the min and max churn
contract_churn["Yes"].max() - contract_churn["Yes"].min()

39.8778190122752

Month-to-month customers have the highest churn rate, while two-year contract customers have the lowest, with a 39.88 percentage-point gap between them. 

Insight: Customers on longer-term contracts show substantially lower churn rates than month-to-month customers

### Tenure and churn

To investigate the customer lifecycle. Do churned customers actually have a lower average tenure than customers who stayed?

In [30]:
df.groupby("Churn")["tenure"].mean()

Churn
No     37.569965
Yes    17.979133
Name: tenure, dtype: float64

Insight #3 — Tenure and churn

Customers who churned had an average tenure of 17.98 months, compared with 37.57 months for customers who stayed — a difference of approximately 19.59 months.

In [31]:
df["tenure_group"] = pd.cut(
    df["tenure"],
    bins=[-1, 12, 24, 48, 72],
    labels=["0–12 months", "13–24 months", "25–48 months", "49–72 months"]
)

In [32]:
tenure_churn = pd.crosstab(
    df["tenure_group"],
    df["Churn"],
    normalize="index"
) * 100

tenure_churn

Churn,No,Yes
tenure_group,,
0–12 months,52.561757,47.438243
13–24 months,71.289062,28.710938
25–48 months,79.611041,20.388959
49–72 months,90.486824,9.513176


Insight #4 — Churn is concentrated among newer customers

Customers in their first 12 months have the highest churn rate, while customers with 49–72 months of tenure have the lowest. Churn declines as customer tenure increases.

The first year is the critical period for customer retention.

In [33]:
### Payment Method vs churn

In [34]:
payment_churn = pd.crosstab(
    df["PaymentMethod"],
    df["Churn"],
    normalize="index"
) * 100

payment_churn.round(2)

Churn,No,Yes
PaymentMethod,,
Bank transfer (automatic),83.29,16.71
Credit card (automatic),84.76,15.24
Electronic check,54.71,45.29
Mailed check,80.89,19.11


Insight #5 — Payment method and churn

Electronic check customers have the highest churn rate at 45.29%, while customers using automatic credit-card payments have the lowest at 15.24% — a 30.05 percentage point gap. 

Customers using electronic checks are substantially more likely to churn than customers using automatic credit-card payments.

In [35]:
### Internet Service VS Churn

In [36]:
internet_churn = pd.crosstab(
    df["InternetService"],
    df["Churn"],
    normalize="index"
) * 100

internet_churn.round(2)

Churn,No,Yes
InternetService,,
DSL,81.04,18.96
Fiber optic,58.11,41.89
No,92.60,7.40


Fiber-optic customers have the highest churn rate at 41.89%, compared with 7.40% among customers without internet service — a 34.49 percentage-point difference. This indicates that fiber-optic customers represent a particularly high-risk segment requiring further investigation.

In [39]:
#To save the cleaned dataset
df.to_csv("customer_churn_cleaned.csv", index=False)

In [40]:
#To verify the cleaned dataset as saved
import os
os.path.exists("customer_churn_cleaned.csv")

True

In [41]:
df["TotalCharges"].dtype

dtype('float64')

In [42]:
df.isna().sum().sum()

np.int64(0)